desc/compute/bounce_integral.py

```diff
@@ -0,0 +1,961 @@
+    kh = jnp.linspace(-t_max, t_max, resolution)
+    h = 2 * t_max / (resolution - 1)
+    x = jnp.tanh(0.5 * jnp.pi * jnp.sinh(kh))
+    w = 0.5 * jnp.pi * h * jnp.cosh(kh) / jnp.cosh(0.5 * jnp.pi * jnp.sinh(kh)) ** 2
+    return x, w
+
+
+_interp1d_vec = jnp.vectorize(
+    interp1d,
+    signature="(m),(n),(n)->(m)",
+    excluded={"method", "derivative", "extrap", "period"},
+)
+
+
+@partial(
+    jnp.vectorize,
+    signature="(m),(n),(n),(n)->(m)",
+    excluded={"method", "derivative", "extrap", "period"},
+)
+def _interp1d_vec_with_df(
+    xq,
+    x,
+    f,
+    fx,
+    method="cubic",
+    derivative=0,
+    extrap=False,
+    period=None,
+):
+    return interp1d(xq, x, f, method, derivative, extrap, period, fx=fx)
+
+
+def _bounce_quad(X, w, knots, B_sup_z, B, B_z_ra, pitch, f, f_method):
+    """Compute bounce quadrature for every pitch along every field line.
+
+    Parameters
+    ----------
+    X : Array, shape(P, S, X.shape[2], w.size)
+        Quadrature points.
+    w : Array, shape(w.size, )
+        Quadrature weights.
+    knots : Array, shape(knots.size, )
+        Field line-following ζ coordinates of spline knots.
+    B_sup_z : Array, shape(S, knots.size, )
+        Contravariant field-line following toroidal component of magnetic field.
+    B : Array, shape(S, knots.size, )
+        Norm of magnetic field.
+    B_z_ra : Array, shape(S, knots.size, )
+        Norm of magnetic field derivative with respect to field-line following label.
+    pitch : Array, shape(P, S)
+        λ values.
+    f : Array, shape(P, S, knots.size, )
+        Function to compute bounce integral of, evaluated at knots.
+    f_method : str
+        Method of interpolation for f.
+
+    Returns
+    -------
+    inner_product : Array, shape(X.shape[:-1])
+        Bounce quadrature for every pitch along every field line.
+
+    """
+    assert pitch.ndim == 2
+    assert w.ndim == knots.ndim == 1
+    assert X.shape == (pitch.shape[0], B.shape[0], X.shape[2], w.size)
+    assert knots.size == B.shape[-1]
+    assert f.ndim == 3 and f.shape[0] == 1 or f.shape[0] == pitch.shape[0]
+    assert f.shape[1:] == B_sup_z.shape == B.shape == B_z_ra.shape
+    # Spline the integrand so that we can evaluate it at quadrature points
+    # without expensive coordinate mappings and root finding.
+    # Spline each function separately so that the singularity near the bounce
+    # points can be captured more accurately than can be by any polynomial.
+    shape = X.shape
+    X = X.reshape(X.shape[0], X.shape[1], -1)
+    if f_method == "constant":
+        f = f[..., 0, jnp.newaxis, jnp.newaxis]
+    else:
+        f = _interp1d_vec(X, knots, f, method=f_method).reshape(shape)
+    # Use akima spline to suppress oscillation.
+    B_sup_z = _interp1d_vec(X, knots, B_sup_z, method="akima").reshape(shape)
+    # Specify derivative at knots for ≈ cubic hermite interpolation.
+    B = _interp1d_vec_with_df(X, knots, B, B_z_ra, method="cubic").reshape(shape)
```

Comment:
Referencing this: https://github.com/PlasmaControl/DESC/pull/854#issuecomment-2046031265,
just change `method="cubic"` to `"cubic"` so that we pass in the argument without specifying it as a keyword.